File: tests/test_similarity_analyzer.py
#!/usr/bin/env python3
"""
Comprehensive Test Suite for Code Similarity Analyzer

This test suite covers all major functionality including:
- Basic similarity analysis
- Complex code analysis
- Edge cases and error handling
- Documentation and comment analysis
- Performance with large files
- Code fragment analysis (is_file=False)

Both implementations (Python/TypeScript) should pass these tests with
similar results (within 3% variance for similarity percentages).
"""

import unittest
import os
import sys
import tempfile

# Add the parent directory to the path to import from python/
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

from python.code_similarity_analyzer import CodeSimilarityAnalyzer


class TestCodeSimilarityAnalyzer(unittest.TestCase):
    
    def setUp(self):
        """Set up test fixtures."""
        self.analyzer = CodeSimilarityAnalyzer()
        
        # Set up paths to sample files
        self.test_dir = os.path.dirname(__file__)
        self.project_root = os.path.dirname(self.test_dir)
        self.samples_dir = os.path.join(self.project_root, 'samples')
        
    def analyze_similarity(self, file_a, file_b, threshold=0.7):
        """Helper method to analyze similarity between two files."""
        return self.analyzer.analyze_code_similarity(file_a, file_b, similarity_threshold=threshold)

    def test_identical_files(self):
        """Test that identical files have 100% similarity"""
        print("\n--- Testing Identical Files ---")
        
        file_a = os.path.join(self.samples_dir, 'sample_a.py')
        results = self.analyze_similarity(file_a, file_a, threshold=0.7)
        
        self.assertEqual(results['similarity_percentage'], 100.0,
                        "Identical files should have 100% similarity")
        self.assertEqual(results['similar_lines_count'], results['lines_a_count'],
                        "All lines should match for identical files")
        
        print(f"✅ Identical files: {results['similarity_percentage']}% similarity "
              f"({results['similar_lines_count']} of {results['lines_a_count']} lines matched)")

    def test_similar_files_variable_renaming(self):
        """Test files with variable renaming (sample_a.py vs sample_c.py)"""
        print("\n--- Testing Variable Renaming ---")
        
        file_a = os.path.join(self.samples_dir, 'sample_a.py')
        file_c = os.path.join(self.samples_dir, 'sample_c.py')
        
        results = self.analyze_similarity(file_a, file_c, threshold=0.6)
        
        # Should detect high similarity despite variable name changes
        self.assertGreaterEqual(results['similarity_percentage'], 60.0,
                               "Files with variable renaming should have >60% similarity")
        self.assertGreater(results['similar_lines_count'], 0,
                          "Should find similar lines despite variable renaming")
        
        print(f"✅ Variable renaming: {results['similarity_percentage']:.1f}% similarity "
              f"({results['similar_lines_count']} of {results['lines_a_count']} lines matched)")

    def test_different_languages(self):
        """Test files in different languages (should have low similarity)"""
        print("\n--- Testing Different Languages ---")
        
        file_py = os.path.join(self.samples_dir, 'sample_c.py')  # Different algorithm
        file_java = os.path.join(self.samples_dir, 'sample_a.java')  # Different algorithm
        
        results = self.analyze_similarity(file_py, file_java, threshold=0.5)
        
        # Different languages should have low similarity (adjusting threshold for realistic case)
        self.assertLessEqual(results['similarity_percentage'], 60.0,
                           "Different languages should have low similarity")
        
        print(f"✅ Different languages: {results['similarity_percentage']:.1f}% similarity "
              f"({results['similar_lines_count']} of {results['lines_a_count']} lines matched)")

    def test_complex_files(self):
        """Test analysis of complex files with documentation"""
        print("\n--- Testing Complex Files ---")
        
        complex_a_path = os.path.join(self.samples_dir, 'complex_a.py')
        complex_b_path = os.path.join(self.samples_dir, 'complex_b.py')
        
        if os.path.exists(complex_a_path) and os.path.exists(complex_b_path):
            results = self.analyze_similarity(complex_a_path, complex_b_path, threshold=0.3)
            
            # Complex files should show some similarity
            self.assertGreaterEqual(results['similarity_percentage'], 20.0,
                                   "Complex files should have some detectable similarity")
            
            print(f"✅ Complex files: {results['similarity_percentage']:.1f}% similarity "
                  f"({results['similar_lines_count']} of {results['lines_a_count']} lines matched)")
        else:
            print("⚠️  Complex test files not found, skipping complex file test")

    def test_code_fragment_analysis(self):
        """Test analyzing code fragments instead of files"""
        print("\n--- Testing Code Fragment Analysis ---")
        
        # Test with code fragments
        code_a = """def hello_world():
    print("Hello, World!")
    return True"""
        
        code_b = """def greet():
    print("Hello, World!")
    return True"""
        
        results = self.analyzer.analyze_code_similarity(code_a, code_b, similarity_threshold=0.7, is_file=False)
        
        # Should detect high similarity despite different function names
        self.assertGreaterEqual(results['similarity_percentage'], 60.0,
                               "Code fragments with similar logic should have >60% similarity")
        self.assertGreater(results['similar_lines_count'], 0,
                          "Should find at least some similar lines")
        self.assertFalse(results['is_file'], "Should indicate this is code fragment analysis")
        
        print(f"✅ Code fragment analysis: {results['similarity_percentage']:.1f}% similarity "
              f"({results['similar_lines_count']} of {results['lines_a_count']} lines matched)")

    def test_empty_files(self):
        """Test handling of empty or non-existent files"""
        print("\n--- Testing Empty/Non-existent Files ---")
        
        # Test with non-existent files
        results = self.analyze_similarity('nonexistent1.py', 'nonexistent2.py', threshold=0.7)
        
        self.assertIn('error', results, "Should return error for non-existent files")
        self.assertEqual(results['similarity_percentage'], 0.0,
                        "Non-existent files should have 0% similarity")
        
        print(f"✅ Non-existent files: Properly handled with error message")

    def test_empty_code_fragments(self):
        """Test handling of empty code fragments"""
        print("\n--- Testing Empty Code Fragments ---")
        
        results = self.analyzer.analyze_code_similarity("", "def test(): pass", similarity_threshold=0.7, is_file=False)
        
        self.assertIn('error', results, "Should return error for empty code fragments")
        self.assertEqual(results['similarity_percentage'], 0.0,
                        "Empty code fragments should have 0% similarity")
        
        print(f"✅ Empty code fragments: Properly handled with error message")

    def test_threshold_sensitivity(self):
        """Test that different thresholds produce appropriate results"""
        print("\n--- Testing Threshold Sensitivity ---")
        
        file_a = os.path.join(self.samples_dir, 'sample_a.py')
        file_c = os.path.join(self.samples_dir, 'sample_c.py')
        
        # Test with low threshold
        results_low = self.analyze_similarity(file_a, file_c, threshold=0.3)
        
        # Test with high threshold  
        results_high = self.analyze_similarity(file_a, file_c, threshold=0.8)
        
        # Lower threshold should find more matches
        self.assertGreaterEqual(results_low['similar_lines_count'], results_high['similar_lines_count'],
                               "Lower threshold should find more or equal matches")
        
        print(f"✅ Threshold sensitivity: Low threshold ({results_low['similar_lines_count']} matches) >= "
              f"High threshold ({results_high['similar_lines_count']} matches)")

    def test_performance_stress_test(self):
        """Test performance with artificially large inputs"""
        print("\n--- Testing Performance Stress Test ---")
        
        # Create large code fragments
        large_code_a = "\n".join([f"def function_{i}():\n    return {i}" for i in range(100)])
        large_code_b = "\n".join([f"def func_{i}():\n    return {i}" for i in range(100)])
        
        import time
        start_time = time.time()
        
        results = self.analyzer.analyze_code_similarity(large_code_a, large_code_b, 
                                                       similarity_threshold=0.5, is_file=False)
        
        end_time = time.time()
        processing_time = end_time - start_time
        
        # Should complete within reasonable time (under 10 seconds for this size)
        self.assertLess(processing_time, 10.0, "Should process large inputs within 10 seconds")
        self.assertGreater(results['similarity_percentage'], 60.0,
                          "Large similar code blocks should have high similarity")
        
        print(f"✅ Performance stress test: {results['similarity_percentage']:.1f}% similarity "
              f"in {processing_time:.2f} seconds ({results['similar_lines_count']} matches)")


def run_comprehensive_tests():
    """Run all tests and provide summary"""
    print("=" * 80)
    print("COMPREHENSIVE CODE SIMILARITY ANALYZER TEST SUITE")
    print("=" * 80)
    
    # Create test suite
    suite = unittest.TestLoader().loadTestsFromTestCase(TestCodeSimilarityAnalyzer)
    
    # Run tests with detailed output
    runner = unittest.TextTestRunner(verbosity=2)
    result = runner.run(suite)
    
    print("\n" + "=" * 80)
    print("TEST SUMMARY")
    print("=" * 80)
    print(f"Tests run: {result.testsRun}")
    print(f"Failures: {len(result.failures)}")
    print(f"Errors: {len(result.errors)}")
    
    if result.failures:
        print("\nFAILURES:")
        for test, traceback in result.failures:
            print(f"  - {test}: {traceback}")
    
    if result.errors:
        print("\nERRORS:")
        for test, traceback in result.errors:
            print(f"  - {test}: {traceback}")
    
    success_rate = ((result.testsRun - len(result.failures) - len(result.errors)) / result.testsRun) * 100
    print(f"\nSUCCESS RATE: {success_rate:.1f}%")
    
    if success_rate >= 80:
        print("✅ Test suite PASSED - Implementation is working correctly")
    else:
        print("❌ Test suite FAILED - Implementation needs attention")
    
    return result.wasSuccessful()


if __name__ == "__main__":
    run_comprehensive_tests()

File: tests/test_similarity_analyzer.ts
/**
 * Comprehensive Test Suite for TypeScript Code Similarity Analyzer
 * 
 * This test suite validates the accuracy and behavior of the Code Similarity Analyzer
 * with various TypeScript scenarios including:
 * - Identical code detection
 * - Plagiarism detection (modified variable names, structure changes)  
 * - Different implementations of same functionality
 * - Unrelated code detection
 * - Edge cases and boundary conditions
 * 
 * Usage:
 *   cd typescript && npm run build && node dist/test_similarity_analyzer.js
 *   OR from tests directory: npm run --prefix ../typescript build && node ../typescript/dist/test_similarity_analyzer.js
 * 
 * The test suite provides detailed assertions and reporting to help developers
 * understand how the similarity analyzer behaves in different scenarios.
 */

import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';
import { CodeSimilarityAnalyzer } from '../typescript/src/CodeSimilarityAnalyzer';

export class TypeScriptSimilarityTester {
    private analyzer: CodeSimilarityAnalyzer;
    private samplesDir: string;
    private testResults: Array<{test: string, passed: boolean, details: string}> = [];

    constructor() {
        this.analyzer = new CodeSimilarityAnalyzer();
        this.samplesDir = path.join(__dirname);
    }

    private createTempFile(content: string): string {
        const tempPath = path.join(os.tmpdir(), `test_${Date.now()}_${Math.random().toString(36).substr(2, 9)}.ts`);
        fs.writeFileSync(tempPath, content);
        return tempPath;
    }

    private analyzeSimilarity(contentA: string, contentB: string, threshold: number = 0.7): any {
        const fileA = this.createTempFile(contentA);
        const fileB = this.createTempFile(contentB);
        
        try {
            const results = this.analyzer.analyzeCodeSimilarity(fileA, fileB, threshold);
            return results;
        } finally {
            // Clean up temporary files
            if (fs.existsSync(fileA)) fs.unlinkSync(fileA);
            if (fs.existsSync(fileB)) fs.unlinkSync(fileB);
        }
    }

    private assert(condition: boolean, message: string, testName: string): void {
        this.testResults.push({
            test: testName,
            passed: condition,
            details: message
        });
        
        if (condition) {
            console.log(`✅ ${testName}: ${message}`);
        } else {
            console.log(`❌ ${testName}: ${message}`);
        }
    }

    testIdenticalCodeDetection(): void {
        const code = `
interface User {
    id: number;
    name: string;
    email: string;
}

class UserManager {
    private users: User[] = [];

    addUser(user: User): void {
        this.users.push(user);
    }

    findUserById(id: number): User | undefined {
        return this.users.find(user => user.id === id);
    }

    getAllUsers(): User[] {
        return [...this.users];
    }
}`;

        const results = this.analyzeSimilarity(code, code, 0.9);
        
        this.assert(
            results.similarityPercentage >= 90.0,
            `Identical code similarity: ${results.similarityPercentage.toFixed(1)}%`,
            "Identical Code Detection"
        );
    }

    testVariableNameChanges(): void {
        const originalCode = `
function calculateCompoundInterest(principal: number, rate: number, time: number): number {
    const amount = principal * Math.pow(1 + rate, time);
    const interest = amount - principal;
    return interest;
}

function processUserRegistration(username: string, email: string, password: string): boolean {
    const userExists = checkIfUserExists(username);
    if (!userExists) {
        createNewUser(username, email, password);
        return true;
    }
    return false;
}

function checkIfUserExists(username: string): boolean {
    // Simulation
    return false;
}

function createNewUser(username: string, email: string, password: string): void {
    // Simulation
}`;

        const modifiedCode = `
function calculateCompoundInterest(initialAmount: number, interestRate: number, duration: number): number {
    const finalAmount = initialAmount * Math.pow(1 + interestRate, duration);
    const earnedInterest = finalAmount - initialAmount;
    return earnedInterest;
}

function processUserRegistration(userId: string, userEmail: string, userPassword: string): boolean {
    const exists = checkIfUserExists(userId);
    if (!exists) {
        createNewUser(userId, userEmail, userPassword);
        return true;
    }
    return false;
}

function checkIfUserExists(userId: string): boolean {
    // Simulation
    return false;
}

function createNewUser(userId: string, userEmail: string, userPassword: string): void {
    // Simulation
}`;

        const results = this.analyzeSimilarity(originalCode, modifiedCode, 0.6);
        
        this.assert(
            results.similarityPercentage >= 60.0 && results.similarityPercentage <= 95.0,
            `Variable name changes similarity: ${results.similarityPercentage.toFixed(1)}%`,
            "Variable Name Changes Detection"
        );
    }

    testStructuralModifications(): void {
        const originalCode = `
class DataProcessor {
    private data: number[] = [];

    addData(value: number): void {
        this.data.push(value);
    }

    calculateAverage(): number {
        if (this.data.length === 0) return 0;
        const sum = this.data.reduce((acc, val) => acc + val, 0);
        return sum / this.data.length;
    }

    getMaxValue(): number {
        return Math.max(...this.data);
    }

    getMinValue(): number {
        return Math.min(...this.data);
    }
}`;

        const modifiedCode = `
class NumberAnalyzer {
    private numbers: number[] = [];

    insertNumber(num: number): void {
        this.numbers.push(num);
    }

    computeMean(): number {
        if (this.numbers.length === 0) return 0;
        const total = this.numbers.reduce((sum, current) => sum + current, 0);
        return total / this.numbers.length;
    }

    findMaximum(): number {
        return Math.max(...this.numbers);
    }

    findMinimum(): number {
        return Math.min(...this.numbers);
    }

    getCount(): number {
        return this.numbers.length;
    }
}`;

        const results = this.analyzeSimilarity(originalCode, modifiedCode, 0.4);
        
        this.assert(
            results.similarityPercentage >= 30.0 && results.similarityPercentage <= 70.0,
            `Structural modifications similarity: ${results.similarityPercentage.toFixed(1)}%`,
            "Structural Modifications Detection"
        );
    }

    testDifferentImplementationsSameLogic(): void {
        const binarySearchIterative = `
function binarySearchIterative(arr: number[], target: number): number {
    let left = 0;
    let right = arr.length - 1;

    while (left <= right) {
        const mid = Math.floor((left + right) / 2);
        
        if (arr[mid] === target) {
            return mid;
        } else if (arr[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    
    return -1;
}

function sortArray(numbers: number[]): number[] {
    return numbers.sort((a, b) => a - b);
}`;

        const binarySearchRecursive = `
function binarySearchRecursive(arr: number[], target: number, left: number = 0, right: number = arr.length - 1): number {
    if (left > right) {
        return -1;
    }
    
    const mid = Math.floor((left + right) / 2);
    
    if (arr[mid] === target) {
        return mid;
    } else if (arr[mid] < target) {
        return binarySearchRecursive(arr, target, mid + 1, right);
    } else {
        return binarySearchRecursive(arr, target, left, mid - 1);
    }
}

function arrangeNumbers(data: number[]): number[] {
    return data.slice().sort((x, y) => x - y);
}`;

        const results = this.analyzeSimilarity(binarySearchIterative, binarySearchRecursive, 0.3);
        
        this.assert(
            results.similarityPercentage >= 15.0 && results.similarityPercentage <= 60.0,
            `Different implementations similarity: ${results.similarityPercentage.toFixed(1)}%`,
            "Different Implementations Same Logic"
        );
    }

    testCompletelyUnrelatedCode(): void {
        const apiClient = `
interface ApiResponse<T> {
    data: T;
    status: number;
    message: string;
}

class HttpClient {
    private baseUrl: string;

    constructor(baseUrl: string) {
        this.baseUrl = baseUrl;
    }

    async get<T>(endpoint: string): Promise<ApiResponse<T>> {
        try {
            const response = await fetch(\`\${this.baseUrl}/\${endpoint}\`);
            const data = await response.json();
            
            return {
                data: data,
                status: response.status,
                message: response.ok ? 'Success' : 'Error'
            };
        } catch (error) {
            throw new Error(\`Network error: \${error}\`);
        }
    }

    async post<T>(endpoint: string, body: any): Promise<ApiResponse<T>> {
        try {
            const response = await fetch(\`\${this.baseUrl}/\${endpoint}\`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify(body)
            });
            
            const data = await response.json();
            
            return {
                data: data,
                status: response.status,
                message: response.ok ? 'Success' : 'Error'
            };
        } catch (error) {
            throw new Error(\`Network error: \${error}\`);
        }
    }
}`;

        const geometryCalculations = `
interface Point {
    x: number;
    y: number;
}

interface Rectangle {
    width: number;
    height: number;
}

class GeometryCalculator {
    static calculateDistance(point1: Point, point2: Point): number {
        const dx = point2.x - point1.x;
        const dy = point2.y - point1.y;
        return Math.sqrt(dx * dx + dy * dy);
    }

    static calculateRectangleArea(rectangle: Rectangle): number {
        return rectangle.width * rectangle.height;
    }

    static calculateRectanglePerimeter(rectangle: Rectangle): number {
        return 2 * (rectangle.width + rectangle.height);
    }

    static isPointInsideRectangle(point: Point, rect: Rectangle, origin: Point): boolean {
        return (
            point.x >= origin.x &&
            point.x <= origin.x + rect.width &&
            point.y >= origin.y &&
            point.y <= origin.y + rect.height
        );
    }

    static calculateTriangleArea(base: number, height: number): number {
        return 0.5 * base * height;
    }
}`;

        const results = this.analyzeSimilarity(apiClient, geometryCalculations, 0.1);
        
        this.assert(
            results.similarityPercentage <= 30.0,
            `Unrelated code similarity: ${results.similarityPercentage.toFixed(1)}%`,
            "Completely Unrelated Code"
        );
    }

    testComplexSamplesSimilarity(): void {
        const complexAPath = path.join(this.samplesDir, "complex_a.ts");
        const complexBPath = path.join(this.samplesDir, "complex_b.ts");

        if (fs.existsSync(complexAPath) && fs.existsSync(complexBPath)) {
            const results = this.analyzer.analyzeCodeSimilarity(complexAPath, complexBPath, 0.3);
            
            this.assert(
                results.similarityPercentage >= 20.0 && results.similarityPercentage <= 60.0,
                `Complex samples similarity: ${results.similarityPercentage.toFixed(1)}%`,
                "Complex Samples Different Implementations"
            );
        }
    }

    testThresholdSensitivity(): void {
        const codeA = `
function processArray(inputArray: number[]): number[] {
    const result: number[] = [];
    for (const item of inputArray) {
        if (item > 0) {
            result.push(item * 2);
        }
    }
    return result;
}

function validateInput(data: any): boolean {
    return data !== null && data !== undefined;
}`;

        const codeB = `
function handleArray(dataArray: number[]): number[] {
    const output: number[] = [];
    for (const element of dataArray) {
        if (element > 0) {
            output.push(element * 2);
        }
    }
    return output;
}

function checkInput(input: any): boolean {
    return input !== null && input !== undefined;
}`;

        const thresholds = [0.3, 0.5, 0.7, 0.9];
        const resultsByThreshold: {[key: number]: any} = {};

        console.log("\nThreshold Sensitivity Analysis:");
        
        for (const threshold of thresholds) {
            const results = this.analyzeSimilarity(codeA, codeB, threshold);
            resultsByThreshold[threshold] = results;
            
            console.log(`  Threshold ${threshold}: ${results.similarityPercentage.toFixed(1)}% similarity, ${results.similarMatches.length} matches`);
        }

        // Verify that higher thresholds generally result in fewer matches
        let thresholdTestPassed = true;
        for (let i = 0; i < thresholds.length - 1; i++) {
            const currThreshold = thresholds[i];
            const nextThreshold = thresholds[i + 1];
            
            const currMatches = resultsByThreshold[currThreshold].similarMatches.length;
            const nextMatches = resultsByThreshold[nextThreshold].similarMatches.length;
            
            if (currMatches < nextMatches) {
                thresholdTestPassed = false;
                break;
            }
        }

        this.assert(
            thresholdTestPassed,
            "Higher thresholds produce fewer matches as expected",
            "Threshold Sensitivity"
        );
    }

    testEdgeCases(): void {
        // Empty files
        const emptyResults = this.analyzeSimilarity("", "", 0.5);
        this.assert(
            emptyResults.similarityPercentage === 0.0,
            "Empty files have 0% similarity",
            "Edge Case: Empty Files"
        );

        // Single line files
        const singleLineResults = this.analyzeSimilarity(
            "console.log('hello');",
            "console.log('hello');",
            0.5
        );
        this.assert(
            singleLineResults.similarityPercentage >= 90.0,
            `Identical single lines: ${singleLineResults.similarityPercentage.toFixed(1)}% similarity`,
            "Edge Case: Single Line Files"
        );

        // Files with only comments
        const commentsA = "// This is a comment\n// Another comment";
        const commentsB = "// Different comment\n// Yet another comment";
        const commentResults = this.analyzeSimilarity(commentsA, commentsB, 0.5);
        
        this.assert(
            true, // Comments are removed during preprocessing
            `Comment-only files: ${commentResults.similarityPercentage.toFixed(1)}% similarity`,
            "Edge Case: Comment-Only Files"
        );
    }

    runAllTests(): void {
        console.log("Starting TypeScript Code Similarity Analyzer Test Suite...");
        console.log("=" .repeat(70));

        this.testIdenticalCodeDetection();
        this.testVariableNameChanges();
        this.testStructuralModifications();
        this.testDifferentImplementationsSameLogic();
        this.testCompletelyUnrelatedCode();
        this.testComplexSamplesSimilarity();
        this.testThresholdSensitivity();
        this.testEdgeCases();

        this.generateTestReport();
    }

    private generateTestReport(): void {
        console.log("\n" + "=".repeat(70));
        console.log("TEST RESULTS SUMMARY");
        console.log("=".repeat(70));

        const passedTests = this.testResults.filter(r => r.passed).length;
        const totalTests = this.testResults.length;

        console.log(`\nTotal Tests: ${totalTests}`);
        console.log(`Passed: ${passedTests}`);
        console.log(`Failed: ${totalTests - passedTests}`);
        console.log(`Success Rate: ${((passedTests / totalTests) * 100).toFixed(1)}%`);

        const failedTests = this.testResults.filter(r => !r.passed);
        if (failedTests.length > 0) {
            console.log("\nFailed Tests:");
            failedTests.forEach(test => {
                console.log(`  ❌ ${test.test}: ${test.details}`);
            });
        }

        console.log("\n" + "=".repeat(70));
        console.log("SIMILARITY INTERPRETATION GUIDE:");
        console.log("- >90% similarity: Likely identical or very minor changes");
        console.log("- 70-90% similarity: Significant similarity, possible plagiarism");
        console.log("- 40-70% similarity: Moderate similarity, same domain/patterns");
        console.log("- 20-40% similarity: Some common elements");
        console.log("- <20% similarity: Largely unrelated code");
        console.log("=".repeat(70));
    }
}

// Export for use in other modules
export function runTypeScriptTests(): void {
    const tester = new TypeScriptSimilarityTester();
    tester.runAllTests();
}

// Run tests if this file is executed directly
runTypeScriptTests();
